astropy/cosmology/core.py

```diff
@@ -104,29 +112,27 @@ def __init_subclass__(cls):
         # -------------------
         # Parameters
 
-        # Get parameters that are still Parameters, either in this class or above.
-        parameters = []
-        derived_parameters = []
-        for n in cls.__parameters__:
-            p = getattr(cls, n)
-            if isinstance(p, Parameter):
-                derived_parameters.append(n) if p.derived else parameters.append(n)
-
-        # Add new parameter definitions
-        for n, v in cls.__dict__.items():
-            if n in parameters or n.startswith("_") or not isinstance(v, Parameter):
+        params = {}
+        derived_params = {}
+        for n, v in all_cls_vars(cls).items():
+            if not isinstance(v, Parameter):
                 continue
-            derived_parameters.append(n) if v.derived else parameters.append(n)
-
-        # reorder to match signature
-        ordered = [
-            parameters.pop(parameters.index(n))
+            if v.derived:
+                derived_params[n] = v
+            else:
+                params[n] = v
+
+        # reorder to match signature, placing "unordered" at the end
+        ordered = {
+            n: params.pop(n)
             for n in cls._init_signature.parameters.keys()
-            if n in parameters
-        ]
-        parameters = ordered + parameters  # place "unordered" at the end
-        cls.__parameters__ = tuple(parameters)
-        cls.__all_parameters__ = cls.__parameters__ + tuple(derived_parameters)
+            if n in params
+        }
+        cls._parameters_ = MappingProxyType(ordered | params)
+        cls._parameters_derived_ = MappingProxyType(derived_params)
+        cls._parameters_all = frozenset(
+            chain(cls.parameters.keys(), cls.derived_parameters.keys())
```

Comment:
Can just be `frozenset(cls.parameters.keys()).union(cls.derived_parameters.keys())` - though perhaps what you have is clearer if slower.

Reply:
This seems good!

Reply:
Done.